### PR TITLE
HARP-11531: Look-at bounds for sphere.

### PR DESCRIPTION
**Pull request by atomicsulfate**
Thank you for contributing to harp.gl!

Before requesting a pull request, please remember to check the following documents:
* [contribution guidelines](https://github.com/heremaps/harp.gl/blob/master/CONTRIBUTING.md)
* [coding style](https://github.com/heremaps/harp.gl/blob/master/CODINGSTYLE.md)

If you are adding new functionality we would highly appreciate if you can describe what is the capability you are adding and even better if you can add some examples. Please also remember to add tests for it.

# CI Check

Our bots will check whether your PR can be directly integrated into the mainline. We have some internal integration tests running on the background, our bots will inform you of the next steps and someone from our team will take a look and help if needed!

And please do not forget to sign-off your commit! You can read more about DCO [here](https://julien.ponge.org/blog/developer-certificate-of-origin-versus-contributor-license-agreements/). But, in short, you just need to use `git commit -s` or append `--signoff` when you are committing to the repo.

Happy contributing!

### CODE REVIEW

File: @here/harp-mapview/lib/BoundsGenerator.ts [hunk 1]
```diff
@@ -11,13 +11,71 @@ import {
...
-import { Frustum, Line3, Matrix4, PerspectiveCamera, Plane, Ray, Vector3 } from "three";
+import { assert } from "@here/harp-utils";
+import { Frustum, Line3, Matrix4, PerspectiveCamera, Plane, Ray, Vector2, Vector3 } from "three";
 
 import { TileCorners } from "./geometry/TileGeometryCreator";
+import { CanvasSide, SphereHorizon } from "./SphereHorizon";
 import { MapViewUtils } from "./Utils";
 
-const logger = LoggerManager.instance.create("BoundsGenerator");
+function computeLongitudeSpan(geoStart: GeoCoordinates, geoEnd: GeoCoordinates): number {
```
[the remark this answers is not in the document]
Comment (atomicsulfate) [reply @here/harp-mapview/lib/BoundsGenerator.ts:21]: done

File: @here/harp-mapview/lib/BoundsGenerator.ts [hunk 2]
```diff
@@ -11,13 +11,71 @@ import {
...
+    const minLongitude = Math.min(geoStart.longitude, geoEnd.longitude);
+    const maxLongitude = Math.max(geoStart.longitude, geoEnd.longitude);
+
+    return Math.min(maxLongitude - minLongitude, 360 + minLongitude - maxLongitude);
+}
+
+// Rough, empirical rule to compute the number of divisions needed for a geopolygon edge to keep
+// the deviation from the view bound edge it must follow within acceptable values.
+function computeEdgeDivisionsForSphere(geoStart: GeoCoordinates, geoEnd: GeoCoordinates): number {
```
[the remark this answers is not in the document]
Comment (atomicsulfate) [reply @here/harp-mapview/lib/BoundsGenerator.ts:30]: This is hardly generic or reusable. It's just a rule of thumb that I came up with just testing how straight do the resulting geolines look once projected into camera space.
Only application I can think of is this bound computation, and if there's a new one coming up, probably the same parameters won't work.

File: @here/harp-mapview/lib/BoundsGenerator.ts [hunk 3]
```diff
@@ -11,13 +11,71 @@ import {
...
+
+const ccwCanvasCornersNDC: Array<{ x: number; y: number }> = [
+    { x: -1, y: -1 }, // bottom left
+    { x: 1, y: -1 }, // bottom right
+    { x: 1, y: 1 }, // top right
+    { x: -1, y: 1 } // top left
+];
+
+// Wrap around coordinates for polygons crossing the antimeridian. Needed for sphere projection.
+function wrapCoordinatesAround(coordinates: GeoCoordinates[]) {
```
[the remark this answers is not in the document]
Comment (atomicsulfate) [reply @here/harp-mapview/lib/BoundsGenerator.ts:55]: done

File: @here/harp-mapview/lib/BoundsGenerator.ts [hunk 4]
```diff
@@ -43,93 +101,275 @@ export class BoundsGenerator {
...
-        //!!!!!!!ALTITUDE IS NOT TAKEN INTO ACCOUNT!!!!!!!!!
-        const coordinates: GeoCoordinates[] = [];
+    private createPolygon(
+        coordinates: GeoCoordinates[],
+        sort: boolean,
+        wrapAround: boolean = false
+    ): GeoPolygon | undefined {
+        if (coordinates.length > 2) {
+            if (wrapAround) {
+                wrapCoordinatesAround(coordinates);
```
[the remark this answers is not in the document]
Comment (atomicsulfate) [reply @here/harp-mapview/lib/BoundsGenerator.ts:116]: done

File: @here/harp-mapview/lib/BoundsGenerator.ts [hunk 5]
```diff
@@ -43,93 +101,275 @@ export class BoundsGenerator {
...
 
-                [
-                    new Ray(worldCorners.se, directionEast), // south east ray
-                    new Ray(worldCorners.se, directionWest), // south west ray
-                    new Ray(worldCorners.ne, directionEast), // north east ray
-                    new Ray(worldCorners.ne, directionWest) //  north west ray
-                ].forEach(ray => {
-                    this.addFrustumIntersection(ray, frustum, coordinates);
-                });
+            // There should only be one horizon intersection, if there's 2 take the last one.
```
[the remark this answers is not in the document]
Comment (atomicsulfate) [reply @here/harp-mapview/lib/BoundsGenerator.ts:194]: I saw it happening even when theoretically it shouldn't. I assumed it was due to some precision problem when the starting side corner barely hits the world (whether the corner hits the world is decided doing raycasting, and horizon visibility is a completely separated computation in SphereHorizon). Let me revisit that to confirm.

File: @here/harp-mapview/lib/BoundsGenerator.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (FraukeF) [reply @here/harp-mapview/lib/BoundsGenerator.ts:23]: ah ok, fine then.

File: @here/harp-examples/src/bounds-generation.ts [hunk 6]
```diff
@@ -30,7 +30,8 @@ export namespace BoundsExample {
   <br />  Press "h" to look at the last created Polygon's BoundingBox
   <br />  Press "g" to look at the last created Polygon
   <br />  Press "b" to show the boundingbox of the Polygon
-  <br />  Press "p" to toggle the projection (!!bounds creation for sphere projection is not yet implemented)`;
+  <br />  Press "p" to toggle the projection (!!bounds creation for sphere projection is not yet implemented)
```
Comment (FraukeF) [inline @here/harp-examples/src/bounds-generation.ts:33]: remove the "not yet implemented.." part

Comment (atomicsulfate) [reply @here/harp-examples/src/bounds-generation.ts]: done

File: @here/harp-mapview/lib/BoundsGenerator.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (atomicsulfate) [reply @here/harp-mapview/lib/BoundsGenerator.ts]: So i've confimed what I've said. I've just rewritten the comment to avoid confusion.

**Comment by FraukeF** [comment]
@harpgl-bot retest this please

**Review by FraukeF — APPROVED** [review]

**Comment by atomicsulfate** [comment]
@harpgl-bot retest this please

**Review by FraukeF — APPROVED** [review]

**Pull request merged**